core/src/main/java/org/springframework/security/oauth2/server/authorization/client/RegisteredClient.java

```diff
@@ -36,4 +50,369 @@
+		 */
+		public Builder scope(Collection<String> scope) {
+			if (scope != null && !scope.isEmpty()) {
+				this.scopes = Collections.unmodifiableSet(new LinkedHashSet<>(scope));
+			}
+			return this;
+		}
+
+		/**
+		 * Builds a new {@link RegisteredClient}.
+		 *
+		 * @return a {@link RegisteredClient}
+		 */
+		public RegisteredClient build() {
+			Assert.notEmpty(this.authorizationGrantTypes, "authorizationGrantTypes cannot be empty");
+			if (authorizationGrantTypes.contains(AuthorizationGrantType.AUTHORIZATION_CODE)) {
+				validateAuthorizationCodeGrantType();
+			}
+			if (authorizationGrantTypes.contains(AuthorizationGrantType.CLIENT_CREDENTIALS)) {
+				validateClientCredentialsGrantType();
+			}
+			this.validateScopes();
+			this.validateRedirectUris();
+			return this.create();
+		}
+
+		private RegisteredClient create() {
+			RegisteredClient registeredClient = new RegisteredClient();
+
+			registeredClient.id = this.id;
+			registeredClient.clientId = this.clientId;
+			registeredClient.clientSecret = StringUtils.hasText(this.clientSecret) ? this.clientSecret : "";
+			registeredClient.clientAuthenticationMethods = this.clientAuthenticationMethods;
+			if (CollectionUtils.isEmpty(this.clientAuthenticationMethods) && StringUtils.hasText(this.clientSecret)) {
+				registeredClient.clientAuthenticationMethods = Collections.singleton(ClientAuthenticationMethod.BASIC);
+			}
+			registeredClient.authorizationGrantTypes = this.authorizationGrantTypes;
+			registeredClient.redirectUris = this.redirectUris;
+			registeredClient.scopes = this.scopes;
+
+			return registeredClient;
+		}
+
+		private void validateClientCredentialsGrantType() {
+			Assert.isTrue(this.authorizationGrantTypes.contains(AuthorizationGrantType.CLIENT_CREDENTIALS),
+					() -> "authorizationGrantTypes must contain " +
+							AuthorizationGrantType.CLIENT_CREDENTIALS.getValue());
+			Assert.hasText(this.id, "id cannot be empty");
+			Assert.hasText(this.clientId, "clientId cannot be empty");
+		}
+
+		private void validateAuthorizationCodeGrantType() {
+			Assert.isTrue(this.authorizationGrantTypes.contains(AuthorizationGrantType.AUTHORIZATION_CODE),
+					() -> "authorizationGrantTypes must contain " +
+							AuthorizationGrantType.AUTHORIZATION_CODE.getValue());
+			Assert.hasText(this.id, "id cannot be empty");
+			Assert.hasText(this.clientId, "clientId cannot be empty");
+			Assert.notEmpty(this.redirectUris, "redirectUris cannot be empty");
+		}
+
+		private void validateScopes() {
+			if (CollectionUtils.isEmpty(this.scopes)) {
+				return;
+			}
+
+			for (String scope : this.scopes) {
+				Assert.isTrue(validateScope(scope), "scope \"" + scope + "\" contains invalid characters");
+			}
+		}
+
+		private static boolean validateScope(String scope) {
+			return scope == null ||
+					scope.chars().allMatch(c -> withinTheRangeOf(c, 0x21, 0x21) ||
+							withinTheRangeOf(c, 0x23, 0x5B) ||
+							withinTheRangeOf(c, 0x5D, 0x7E));
+		}
+
```

Comment:
I can extract it out if its needed in other classes as well. 
@jgrandja Please let me know your thoughts as well on this.

Reply:
1. extracting this into a utility class might e.g. be useful for configuration applications. Someone might create an application that is able to dynamically add new scopes and map these on specific user details (this is possible in keycloak for example). 
2. my other thought on this is simply codestyle. This is logic that is directly related to the scopes and therefore the code readability would be enhanced if this was extracted into its own class.
3. there are more operations needed on scopes than just the validating the scopes. I think all operations on scopes should be found in a single place instead of being spread through the whole API.

I would also prefer non static methods in case that for any reason the developer needs to change some of the logic

Reply:
Let's extract logic when it's needed. Right now it's not needed so the preference is to encapsulate for now.